### PR TITLE
feat(Tag): Prevent duplicate tags when editing tag

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes
- Created EditTagDialogComponent
- Created AbstractTagDialogComponent that CreateTagDialogComponent and EditTagDialogComponent both extend
- Prevent duplicate tags when editing a tag

## Test
- Test with https://github.com/WISE-Community/WISE-API/pull/269
- Make sure you can't create a duplicate tag when editing a tag
- Make sure you can't create a duplicate tag when creating a new tag

### CODE REVIEW

File: src/app/teacher/abstract-tag-dialog.component.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import { ProjectTagService } from '../../assets/wise5/services/projectTagService';
+import { Tag } from '../domain/tag';
+
+@Directive()
+export abstract class AbstractTagDialogComponent implements OnInit {
+  protected tagControl = new FormControl('', [
+    Validators.required,
+    this.createUniqueTagValidator()
+  ]);
+  protected tags: Tag[] = [];
```
Comment (hirokiterashima) [inline src/app/teacher/abstract-tag-dialog.component.ts]: If you move the order of these variables, you won't need to check for null in ```createUniqueTagValidator()``` 
```suggestion
    protected tags: Tag[] = [];
    protected tagControl = new FormControl('', [
    Validators.required,
    this.createUniqueTagValidator()
  ]);
```

File: src/app/teacher/abstract-tag-dialog.component.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  ngOnInit(): void {
+    this.projectTagService.retrieveUserTags().subscribe((tags: Tag[]) => {
+      this.tags = tags;
+    });
+  }
+
+  private createUniqueTagValidator(): ValidatorFn {
+    return (control: AbstractControl): ValidationErrors | null => {
+      return this.tags != null && this.doesTagAlreadyExist(this.tags, control.value)
```
Comment (hirokiterashima) [inline src/app/teacher/abstract-tag-dialog.component.ts]: Remove null check and only pass in the new tagText to doesTagAlreadyExist()?
```suggestion
      return this.doesTagAlreadyExist(control.value)
```

File: src/app/teacher/abstract-tag-dialog.component.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        ? { tagAlreadyExists: true }
+        : null;
+    };
+  }
+
+  private doesTagAlreadyExist(tags: Tag[], tagText: string): boolean {
+    return tags.some((tag: Tag) => tag.text.toLowerCase() === tagText.toLowerCase().trim());
+  }
```
Comment (hirokiterashima) [inline src/app/teacher/abstract-tag-dialog.component.ts]: Get tags via ```this.tags```?
```suggestion
  private doesTagAlreadyExist(tagText: string): boolean {
    return this.tags.some((tag: Tag) => tag.text.toLowerCase() === tagText.toLowerCase().trim());
  }
```

File: src/app/teacher/abstract-tag-dialog.component.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  protected keyPressed(event: KeyboardEvent): void {
+    if (event.key === 'Enter') {
+      this.enterKeyAction();
+    }
+  }
+
+  protected abstract enterKeyAction(): void;
```
Comment (hirokiterashima) [inline src/app/teacher/abstract-tag-dialog.component.ts]: You won't need these functions if you change
```
(keypress)="keyPressed($event)"
```
to
```
(keyup.enter)="create()"
```
or 
```
(keyup.enter)="save()"
```
in create-tag-dialog.component.html and in edit-tag-dialog.component.html?
```suggestion
```

File: src/app/teacher/edit-tag-dialog/edit-tag-dialog.component.html [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<h2 mat-dialog-title i18n>Edit Tag</h2>
+<mat-dialog-content>
+  <mat-form-field class="tag-name">
+    <mat-label i18n>Tag Name</mat-label>
+    <input matInput (keypress)="keyPressed($event)" [formControl]="tagControl" autofocus />
```
Comment (hirokiterashima) [inline src/app/teacher/edit-tag-dialog/edit-tag-dialog.component.html]: Call save() directly on keyup.enter event?
```suggestion
    <input matInput (keyup.enter)="save()" [formControl]="tagControl" autofocus />
```

File: src/app/teacher/create-tag-dialog/create-tag-dialog.component.ts [hunk 6]
```diff
@@ -33,53 +24,27 @@ import { Tag } from '../../domain/tag';
...
-  }
-
-  private createUniqueTagValidator(): ValidatorFn {
-    return (control: AbstractControl): ValidationErrors | null => {
-      return this.tags != null &&
-        this.projectTagService.doesTagAlreadyExist(this.tags, control.value)
-        ? { tagAlreadyExists: true }
-        : null;
-    };
+  ngOnInit(): void {
```
Comment (hirokiterashima) [inline src/app/teacher/create-tag-dialog/create-tag-dialog.component.ts]: Remove function?

File: src/app/teacher/create-tag-dialog/create-tag-dialog.component.ts [hunk 7]
```diff
@@ -33,53 +24,27 @@ import { Tag } from '../../domain/tag';
...
+    super.ngOnInit();
   }
 
-  protected keyPressed(event: KeyboardEvent): void {
-    if (event.key === 'Enter') {
-      this.create();
-    }
+  protected enterKeyAction(): void {
```
Comment (hirokiterashima) [inline src/app/teacher/create-tag-dialog/create-tag-dialog.component.ts]: Remove function and call create() directly from template on keyup.enter event?

File: src/app/teacher/edit-tag-dialog/edit-tag-dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  ngOnInit(): void {
+    super.ngOnInit();
+    this.tagControl.setValue(this.tag.text);
+  }
+
+  protected enterKeyAction(): void {
+    this.save();
+  }
```
Comment (hirokiterashima) [inline src/app/teacher/edit-tag-dialog/edit-tag-dialog.component.ts]: Remove function and call save() directly from template on keyup.enter event?
```suggestion
```

**Review by hirokiterashima — COMMENTED** [review]
Works as described. I think the code can be cleaned up a bit more. I added comments inline. Lmk what you think.

**Comment by geoffreykwan** [comment]
Ok, I've made the changes.

**Review by hirokiterashima — APPROVED** [review]
LGTM.

I noticed that validation error message only gets displayed as a mat-error after focus out. The save button is disabled, which is good, but it would be nice to why (the error message) too. Is this something that can be displayed as soon as duplication is detected?

**Pull request merged**